pkg/apis/provisioning/v1alpha5/requirements.go

```diff
@@ -70,126 +76,244 @@ var (
 	}
 )
 
-// Requirements is a decorated alias type for []v1.NodeSelectorRequirements
-type Requirements []v1.NodeSelectorRequirement
-
-func (r Requirements) Zones() sets.String {
-	return r.Requirement(v1.LabelTopologyZone)
-}
-
-func (r Requirements) InstanceTypes() sets.String {
-	return r.Requirement(v1.LabelInstanceTypeStable)
+type Requirements struct {
+	// Requirements are layered with Labels and applied to every node.
+	Requirements []v1.NodeSelectorRequirement `json:"requirements,omitempty"`
+	// Ground truth for record keeping
+	Records map[string][]v1.NodeSelectorRequirement `json:"-"`
+	Allows  map[string]sets.Set                     `json:"-"`
 }
 
-func (r Requirements) Architectures() sets.String {
-	return r.Requirement(v1.LabelArchStable)
+func NewRequirements(requirements ...v1.NodeSelectorRequirement) *Requirements {
+	result := Requirements{
+		Records: map[string][]v1.NodeSelectorRequirement{},
+		Allows:  map[string]sets.Set{},
+	}
+	return result.Add(requirements...)
 }
 
-func (r Requirements) OperatingSystems() sets.String {
-	return r.Requirement(v1.LabelOSStable)
+func (r *Requirements) Add(requirements ...v1.NodeSelectorRequirement) *Requirements {
+	if r == nil {
+		return NewRequirements(requirements...)
+	}
+	rd := r.DeepCopy()
+	for _, requirement := range requirements {
+		// Normalizing the key
+		if newKey, ok := NormalizedLabels[requirement.Key]; ok {
+			requirement.Key = newKey
+		}
+		//rd.Requirements = append(rd.Requirements, requirement)
+		rd.Records[requirement.Key] = append(rd.Records[requirement.Key], requirement)
+		switch requirement.Operator {
+		case v1.NodeSelectorOpIn:
+			rd.Allows[requirement.Key] = rd.Allow(requirement.Key).Intersection(sets.NewSet(requirement.Values...))
+		case v1.NodeSelectorOpNotIn:
+			rd.Allows[requirement.Key] = rd.Allow(requirement.Key).Intersection(sets.NewComplementSet(requirement.Values...))
+		}
+	}
+	return rd
 }
 
-func (r Requirements) CapacityTypes() sets.String {
-	return r.Requirement(LabelCapacityType)
+func (r *Requirements) MarshalJSON() ([]byte, error) {
```

Comment:
I think you only needed to do this because of the bug I pointed out here: https://github.com/aws/karpenter/pull/1155/files#r791073256

Reply:
I was hoping there is a simple explanation. But the way I discovered this special combinations of tags and MarshalJSON functions that finally works can confirm that MarshalJSON is really needed.

So with just a simple `json:",inline"`, I noticed that data sent to the API server is actually an object instead of a slice. That's when I think for some reasons the build-in Marshaling function failed to understand the `json:",inline"` tag structure. Once I added the MarhsalJSON and UnmarsalJSON functions, I noticed that in the API call there are two slices, one under `Requirements` and one under `requirements`. Strange, but then I tried to force it by adding the `json:"requirements,inline". At least now in most of the test cases where requirements with none-nil contents works.

Then I encounter nil pointer error. That's when I found that for empty requirements, the API call still include `requirements` key. That's when I added the `json:"requirements,inline,omitempty"`.

Very strange case. I spent two days stretching my head trying to figure out why the behavior and settle with this abomination.